<a href="https://colab.research.google.com/github/uvspraneeth/data-cleaning-and-the-skies-Data-Camp-competition/blob/master/workbook.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Cleaning data and the skies
> Using `python` for data analysis.

### Importing Libraries

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
!pip install plotly

In [ ]:
import os
import re
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from plotly import express as ex

# Configuration
pd.set_option('display.max_columns', 500)
pd.set_option('display.max_rows', None)

### Loading Data

In [ ]:
data = pd.read_csv("/content/drive/MyDrive/PROJECTS/data camp competition/ozone.csv")

# data info
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54759 entries, 0 to 54758
Data columns (total 17 columns):
 #   Column                                Non-Null Count  Dtype  
---  ------                                --------------  -----  
 0   Date                                  54759 non-null  object 
 1   Source                                54759 non-null  object 
 2   Site ID                               54759 non-null  int64  
 3   POC                                   54759 non-null  int64  
 4   Daily Max 8-hour Ozone Concentration  52021 non-null  float64
 5   Units                                 54759 non-null  object 
 6   Daily AQI Value                       52021 non-null  float64
 7   Local Site Name                       54759 non-null  object 
 8   Daily Obs Count                       54759 non-null  int64  
 9   Percent Complete                      54759 non-null  float64
 10  Method Code                           48269 non-null  float64
 11  CBSA Code      

In [ ]:
# top 5 records of the data
data.head()

Date Source   Site ID  POC  \
0            /2024    AQS  60010007    1   
1       01/02/2024    AQS  60010007    1   
2            /2024    AQS  60010007    1   
3  January 04/2024    AQS  60010007    1   
4  January 05/2024    AQS  60010007    1   

   Daily Max 8-hour Ozone Concentration Units  Daily AQI Value  \
0                                 0.031   ppm             29.0   
1                                 0.037   ppm             34.0   
2                                   NaN   ppm             30.0   
3                                 0.026   ppm             24.0   
4                                 0.027   ppm             25.0   

  Local Site Name  Daily Obs Count  Percent Complete  Method Code  CBSA Code  \
0       Livermore               17             100.0         47.0    41860.0   
1       Livermore               17             100.0         47.0    41860.0   
2       Livermore               17             100.0         47.0    41860.0   
3       Livermore               17             100.0         47.0    41860.0   
4       Livermore               17             100.0         47.0    41860.0   

                           CBSA Name  County FIPS Code   County  \
0  San Francisco-Oakland-Hayward, CA                 1  Alameda   
1  San Francisco-Oakland-Hayward, CA                 1  Alameda   
2  San Francisco-Oakland-Hayward, CA                 1  Alameda   
3  San Francisco-Oakland-Hayward, CA                 1  Alameda   
4  San Francisco-Oakland-Hayward, CA                 1  Alameda   

   Site Latitude  Site Longitude  
0      37.687526     -121.784217  
1      37.687526     -121.784217  
2      37.687526     -121.784217  
3      37.687526     -121.784217  
4      37.687526     -121.784217

We can notice here there exists lots of inconsistency in the data, such as `Date`, `Daily max 8-hour Ozone Concentration` etc. And it looks like time series data.

In [ ]:
data.describe()

Site ID           POC  Daily Max 8-hour Ozone Concentration  \
count  5.475900e+04  54759.000000                          52021.000000   
mean   6.059086e+07      1.000256                              0.043519   
std    3.001943e+05      0.015988                              0.014610   
min    6.001001e+07      1.000000                              0.000000   
25%    6.031100e+07      1.000000                              0.034000   
50%    6.065500e+07      1.000000                              0.041000   
75%    6.081100e+07      1.000000                              0.052000   
max    6.113100e+07      2.000000                              0.139000   

       Daily AQI Value  Daily Obs Count  Percent Complete   Method Code  \
count     52021.000000     54759.000000      54759.000000  48269.000000   
mean         45.280079        27.534688         99.340346     89.051669   
std          25.899065        97.800446          3.658452     36.636778   
min           0.000000         1.000000          6.000000     47.000000   
25%          31.000000        17.000000        100.000000     87.000000   
50%          38.000000        17.000000        100.000000     87.000000   
75%          48.000000        17.000000        100.000000     87.000000   
max         236.000000      1000.000000        100.000000    199.000000   

          CBSA Code  County FIPS Code  Site Latitude  Site Longitude  
count  52351.000000      54759.000000   54759.000000    54759.000000  
mean   36107.577697         58.918406      36.169550     -119.665107  
std     9536.717562         30.012874       2.192841        2.022987  
min    12540.000000          1.000000      32.578165     -124.179490  
25%    31080.000000         31.000000      34.181977     -121.420838  
50%    40140.000000         65.000000      36.209286     -119.828400  
75%    41860.000000         81.000000      37.961578     -118.068500  
max    49700.000000        113.000000      41.726892     -114.602090

In [ ]:
data.duplicated().sum()

np.int64(3576)

In [ ]:
# check for nulls
data.isna().sum()

Date                                       0
Source                                     0
Site ID                                    0
POC                                        0
Daily Max 8-hour Ozone Concentration    2738
Units                                      0
Daily AQI Value                         2738
Local Site Name                            0
Daily Obs Count                            0
Percent Complete                           0
Method Code                             6490
CBSA Code                               2408
CBSA Name                               2408
County FIPS Code                           0
County                                     0
Site Latitude                              0
Site Longitude                             0
dtype: int64

In [ ]:
# Let's drop the duplicates from the data
data.drop_duplicates(inplace=True)

In [ ]:
# verify
data.duplicated().sum()

np.int64(0)

In [ ]:
data.isna().sum()

Date                                       0
Source                                     0
Site ID                                    0
POC                                        0
Daily Max 8-hour Ozone Concentration    2726
Units                                      0
Daily AQI Value                         2722
Local Site Name                            0
Daily Obs Count                            0
Percent Complete                           0
Method Code                             6168
CBSA Code                               2248
CBSA Name                               2248
County FIPS Code                           0
County                                     0
Site Latitude                              0
Site Longitude                             0
dtype: int64

### Type fixing

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 51183 entries, 0 to 54758
Data columns (total 17 columns):
 #   Column                                Non-Null Count  Dtype  
---  ------                                --------------  -----  
 0   Date                                  51183 non-null  object 
 1   Source                                51183 non-null  object 
 2   Site ID                               51183 non-null  int64  
 3   POC                                   51183 non-null  int64  
 4   Daily Max 8-hour Ozone Concentration  48457 non-null  float64
 5   Units                                 51183 non-null  object 
 6   Daily AQI Value                       48461 non-null  float64
 7   Local Site Name                       51183 non-null  object 
 8   Daily Obs Count                       51183 non-null  int64  
 9   Percent Complete                      51183 non-null  float64
 10  Method Code                           45015 non-null  float64
 11  CBSA Code           

In [ ]:
data[(data['CBSA Code'].isna()) & (data['CBSA Name'].notna())].size

0

In [ ]:
data[(data['CBSA Code'].notna()) & (data['CBSA Name'].isna())].size

0

In [ ]:
data.fillna({'CBSA Code': 999}, inplace=True)
data.fillna({'CBSA Name': 'No_CBSA'}, inplace=True)

Filling the `NaN` CBSA Code value with `999` and `NaN` CBSA Name with `No CBSA` to ensure clean data.


In [ ]:
# monitored records from sources count.
data['Source'].value_counts()

Source
AQS       45015
AirNow     6168
Name: count, dtype: int64

In [ ]:
data.loc[(data['Source'] == 'AirNow') & (data['Method Code'].isna()), 'Method Code'].size

6168

In [ ]:
data.loc[(data['Source'] == 'AirNow') & (data['Method Code'].notna()), 'Method Code'].size

0

From the above observation, We can conclude that, The data from source `AirNow` have no method code value, which is cause of `6168` number of records with Method Code `NaN` value.

> We can fix this by imputing/filling the `NaN` in Methode Code with `999` to ensure clean information.

In [ ]:
data.fillna({'Method Code': 999}, inplace=True)

In [ ]:
sites_per_county = data.groupby('County')['Site ID'].nunique().reset_index()

In [ ]:
sites_per_county

County  Site ID
0           Alameda        6
1            Amador        1
2             Butte        2
3         Calaveras        1
4            Colusa        1
5      Contra Costa        4
6         El Dorado        3
7            Fresno        7
8             Glenn        1
9          Humboldt        1
10         Imperial        4
11             Inyo        3
12             Kern        8
13            Kings        1
14               LA       11
15             Lake        1
16      Los Angeles       12
17           Madera        2
18            Marin        1
19         Mariposa        2
20        Mendocino        2
21           Merced        2
22         Monterey        3
23           Nevada        1
24           Orange        2
25           Placer        5
26        Riverside       11
27               SF        1
28       Sacramento        5
29       San Benito        2
30   San Bernardino        9
31        San Diego        9
32    San Francisco        1
33      San Joaquin        2
34  San Luis Obispo        6
35        San Mateo        1
36    Santa Barbara        9
37      Santa Clara        4
38       Santa Cruz        1
39           Shasta        3
40         Siskiyou        1
41           Solano        3
42           Sonoma        1
43       Stanislaus        2
44           Sutter        2
45           Tehama        2
46           Tulare        4
47         Tuolumne        1
48          Ventura        5
49             Yolo        2

In [ ]:
sites_per_county['Site ID'].sum()

np.int64(174)

In [ ]:
sites_per_county['County'].to_list()

['Alameda',
 'Amador',
 'Butte',
 'Calaveras',
 'Colusa',
 'Contra Costa',
 'El Dorado',
 'Fresno',
 'Glenn',
 'Humboldt',
 'Imperial',
 'Inyo',
 'Kern',
 'Kings',
 'LA',
 'Lake',
 'Los Angeles',
 'Madera',
 'Marin',
 'Mariposa',
 'Mendocino',
 'Merced',
 'Monterey',
 'Nevada',
 'Orange',
 'Placer',
 'Riverside',
 'SF',
 'Sacramento',
 'San Benito',
 'San Bernardino',
 'San Diego',
 'San Francisco',
 'San Joaquin',
 'San Luis Obispo',
 'San Mateo',
 'Santa Barbara',
 'Santa Clara',
 'Santa Cruz',
 'Shasta',
 'Siskiyou',
 'Solano',
 'Sonoma',
 'Stanislaus',
 'Sutter',
 'Tehama',
 'Tulare',
 'Tuolumne',
 'Ventura',
 'Yolo']

You can observe of here, we have `LA` and `Los Angeles` similarly `SF` and `San Francisco`.
```
Duplicates (actually same counties)
LA ↔ Los Angeles

Same county, one is abbreviated.

SF ↔ San Francisco

Same county, one is abbreviated.
```

Let's fix this.

In [ ]:
county_map = {
    'LA' : 'Los Angeles',
    'SF' : 'San Francisco'
}
data['County'] = data['County'].replace(county_map)

In [ ]:
data.drop_duplicates(inplace=True)

In [ ]:
data.duplicated().sum()

np.int64(0)

In [ ]:
# Let's change the type of Method Code and CBSA Code to int
data[['CBSA Code', 'Method Code']] = data[['CBSA Code', 'Method Code']].astype(int)

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 51181 entries, 0 to 54758
Data columns (total 17 columns):
 #   Column                                Non-Null Count  Dtype  
---  ------                                --------------  -----  
 0   Date                                  51181 non-null  object 
 1   Source                                51181 non-null  object 
 2   Site ID                               51181 non-null  int64  
 3   POC                                   51181 non-null  int64  
 4   Daily Max 8-hour Ozone Concentration  48455 non-null  float64
 5   Units                                 51181 non-null  object 
 6   Daily AQI Value                       48459 non-null  float64
 7   Local Site Name                       51181 non-null  object 
 8   Daily Obs Count                       51181 non-null  int64  
 9   Percent Complete                      51181 non-null  float64
 10  Method Code                           51181 non-null  int64  
 11  CBSA Code           

Now evrthing looks better except the `Daily Max 8-hour Ozone Concentration`, `Daily AQI Value`, and `Date`.
we'll fix 'em in the next section.
[Cleaning](#cleaning-the-data)

### Cleaning the data

In [ ]:
df = data.copy()

In [ ]:
def clean_dates(df, date_col, new_col = 'clean_date'):
  series = df[date_col].astype(str).str.strip()

  junk = series.str.match(r'^[/\s]*2024.*|^[/\s]*$')
  remaining = ~junk

  cleaned_dates = pd.Series(pd.NaT, index=series.index)

  # Standard parse
  try_parsed = pd.to_datetime(series[remaining], errors='coerce')
  cleaned_dates.loc[remaining] = try_parsed

  # Fix weird/mixed formats
  def clean_weirdos(date_str):
    try:
      match = re.match(r'^([A-Za-z]+)\s+(\d{1,2})/(\d{4})', date_str)
      if match:
        month, day, year = match.groups()
        return pd.to_datetime(f'{day} {month} {year}', format= '%d %B %Y', errors='coerce')
      return pd.NaT
    except:
      return pd.NaT

  weird_dates = (remaining) & (cleaned_dates.isna())
  cleaned_dates.loc[weird_dates] = series[weird_dates].apply(clean_weirdos)

  # Update DataFrame
  df[new_col] = cleaned_dates

  while True:
    null_count = df[new_col].isna().sum()
    if null_count == 0:
      break

    # replace the missing dates with Lag/Lead fills
    df['lag'] = df[new_col].shift(1)
    df['lead'] = df[new_col].shift(-1)

    lag_mask = df[new_col].isna() &  df['lag'].notna()
    df.loc[lag_mask, new_col] = df.loc[lag_mask, 'lag'] + pd.Timedelta(days=1)
    lead_mask = df[new_col].isna() & df['lead'].notna()
    df.loc[lead_mask, new_col] = df.loc[lead_mask, 'lead'] - pd.Timedelta(days=1)

    # If nothing changed this round -> break (avoid infinite loops)
    if df[new_col].isna().sum() == null_count:
        break
  df[date_col] = df[new_col]
  df.drop(columns=['lag', 'lead', new_col], inplace=True)
  return df

In [ ]:
df = df.groupby('Site ID', group_keys=False).apply(lambda g: clean_dates(g, 'Date'))

/tmp/ipython-input-979613763.py:1: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  df = df.groupby('Site ID', group_keys=False).apply(lambda g: clean_dates(g, 'Date'))


In [ ]:
df.head()

Date Source   Site ID  POC  Daily Max 8-hour Ozone Concentration  \
0 2024-01-01    AQS  60010007    1                                 0.031   
1 2024-01-02    AQS  60010007    1                                 0.037   
2 2024-01-03    AQS  60010007    1                                   NaN   
3 2024-01-04    AQS  60010007    1                                 0.026   
4 2024-01-05    AQS  60010007    1                                 0.027   

  Units  Daily AQI Value Local Site Name  Daily Obs Count  Percent Complete  \
0   ppm             29.0       Livermore               17             100.0   
1   ppm             34.0       Livermore               17             100.0   
2   ppm             30.0       Livermore               17             100.0   
3   ppm             24.0       Livermore               17             100.0   
4   ppm             25.0       Livermore               17             100.0   

   Method Code  CBSA Code                          CBSA Name  \
0           47      41860  San Francisco-Oakland-Hayward, CA   
1           47      41860  San Francisco-Oakland-Hayward, CA   
2           47      41860  San Francisco-Oakland-Hayward, CA   
3           47      41860  San Francisco-Oakland-Hayward, CA   
4           47      41860  San Francisco-Oakland-Hayward, CA   

   County FIPS Code   County  Site Latitude  Site Longitude  
0                 1  Alameda      37.687526     -121.784217  
1                 1  Alameda      37.687526     -121.784217  
2                 1  Alameda      37.687526     -121.784217  
3                 1  Alameda      37.687526     -121.784217  
4                 1  Alameda      37.687526     -121.784217

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 51181 entries, 0 to 54758
Data columns (total 17 columns):
 #   Column                                Non-Null Count  Dtype         
---  ------                                --------------  -----         
 0   Date                                  51181 non-null  datetime64[ns]
 1   Source                                51181 non-null  object        
 2   Site ID                               51181 non-null  int64         
 3   POC                                   51181 non-null  int64         
 4   Daily Max 8-hour Ozone Concentration  48455 non-null  float64       
 5   Units                                 51181 non-null  object        
 6   Daily AQI Value                       48459 non-null  float64       
 7   Local Site Name                       51181 non-null  object        
 8   Daily Obs Count                       51181 non-null  int64         
 9   Percent Complete                      51181 non-null  float64       
 10  Met

Now let's clean `Daily Max 8-hour Ozone Concentration` and `Daily AQI Value`, Since these both are interelated features, we fix 'em by calculating.

**AQI Calculation: How Does It Work?** \
**Understanding AQI Formulas** \
The Air Quality Index (AQI) is calculated using a standardized formula that converts the concentration of individual pollutants into a single numerical value. Each pollutant has its own sub-index, and the highest sub-index value determines the overall AQI.

The general formula for calculating the AQI for a specific pollutant is:

$$ AQI = \frac{I_{high} - I_{low}}{C_{high} - C_{low}} \times (C - C_{low}) + I_{low} $$
\
Where:

$ AQI $: Air Quality Index value for the specific pollutant (**Ozone $O_3$**) \
$C$: Measured pollutant concentration. (**In our case Ozone $O_3$ Concentration.**) \
$C_{high} , C_{low}$: Breakpoints (*lower and upper concentration limits for the pollutant's AQI category*)
\
$I_{high}, I_{low}$: AQI values corresponding to the breakpoints

[reference](https://www.freeonlinecalc.com/air-quality-index-aqi-calculation-review-and-formulas.html)

**For $Ozone (O_3)$ Calculation:** \
we can derive from the **AQI Calculation**, \

$$ C (O_3 \space concentration) = \frac{C_{high} - C_{low}}{I_{high} - I_{low}} \times (AQI - I_{low}) + C_{low}$$

**Pollutant Concentration to AQI Conversion**

| AQI Category                  | AQI Range | Ozone (O3) 8-hour (ppm) |
| :---------------------------- | :-------- | :---------------------- |
| Good                          | 0 - 50    | 0.000 - 0.054           |
| Moderate                      | 51 - 100  | 0.055 - 0.070           |
| Unhealthy for Sensitive Groups| 101 - 150 | 0.071 - 0.085           |
| Unhealthy                     | 151 - 200 | 0.086 - 0.105           |
| Very Unhealthy                | 201 - 300 | 0.106 - 0.200           |
| Hazardous                     | 301 - 500 | 0.201 - 0.600           |

In [ ]:
def calculate_aqi_and_ozone(aqi = None, ozone = None, flag = False):
  benchmark = [
      (0, 50, 0.000, 0.054),
      (51, 100, 0.055, 0.070),
      (101, 150,  0.071, 0.085),
      (151, 200, 0.086, 0.105),
      (201, 300, 0.106, 0.200),
      (301, 500, 0.201, 0.600)
  ]

  for i_low, i_high, c_low, c_high in benchmark:
    if (flag) and (i_low <= aqi and aqi <= i_high):
      return ((c_high - c_low)/(i_high - i_low)) * (aqi - i_low) + c_high
    elif (not flag) and (c_low <= ozone and ozone <= c_high):
      return ((i_high - i_low) / (c_high - c_low)) * (ozone - c_low) + i_low

In [ ]:
# Filling the missing `Daily AQI Value` when `Daily Max 8-hour Ozone Concentration` value exists
aqi_mask = df['Daily AQI Value'].isna() & df['Daily Max 8-hour Ozone Concentration'].notna()
df.loc[aqi_mask, 'Daily AQI Value'] = df.loc[aqi_mask, 'Daily Max 8-hour Ozone Concentration'].apply(lambda o: calculate_aqi_and_ozone(ozone=o))

# Filling the `missing Daily Max 8-hour Ozone Concentration` value when `Daily AQI Value` exists
ozone_mask = df['Daily AQI Value'].notna() & df['Daily Max 8-hour Ozone Concentration'].isna()
df.loc[ozone_mask, 'Daily Max 8-hour Ozone Concentration'] = df.loc[ozone_mask, 'Daily AQI Value'].apply(lambda a: calculate_aqi_and_ozone(aqi = a, flag=True))

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 51181 entries, 0 to 54758
Data columns (total 17 columns):
 #   Column                                Non-Null Count  Dtype         
---  ------                                --------------  -----         
 0   Date                                  51181 non-null  datetime64[ns]
 1   Source                                51181 non-null  object        
 2   Site ID                               51181 non-null  int64         
 3   POC                                   51181 non-null  int64         
 4   Daily Max 8-hour Ozone Concentration  51047 non-null  float64       
 5   Units                                 51181 non-null  object        
 6   Daily AQI Value                       51047 non-null  float64       
 7   Local Site Name                       51181 non-null  object        
 8   Daily Obs Count                       51181 non-null  int64         
 9   Percent Complete                      51181 non-null  float64       
 10  Met

In [ ]:
junk = df['Daily AQI Value'].isna() & df['Daily Max 8-hour Ozone Concentration'].isna()
junk.sum()

np.int64(134)

It looks like there are still a few rows where `Daily Max 8-hour Ozone Concentration` and `Daily AQI Value` are missing. We can remove these rows as we can't impute them with other columns. So we drop these rows.

In [ ]:
df.dropna(inplace = True)

In [ ]:
# Finally drop duplicates if exists
df.drop_duplicates(inplace=True)

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 50954 entries, 0 to 54758
Data columns (total 17 columns):
 #   Column                                Non-Null Count  Dtype         
---  ------                                --------------  -----         
 0   Date                                  50954 non-null  datetime64[ns]
 1   Source                                50954 non-null  object        
 2   Site ID                               50954 non-null  int64         
 3   POC                                   50954 non-null  int64         
 4   Daily Max 8-hour Ozone Concentration  50954 non-null  float64       
 5   Units                                 50954 non-null  object        
 6   Daily AQI Value                       50954 non-null  float64       
 7   Local Site Name                       50954 non-null  object        
 8   Daily Obs Count                       50954 non-null  int64         
 9   Percent Complete                      50954 non-null  float64       
 10  Met

Let's save the data for further process.
data is saved as `cleaned_ozone.csv`

In [ ]:
# Saving the cleaned data frame to cleaned_ozone.csv file
df.to_csv('/content/drive/MyDrive/PROJECTS/data camp competition/cleaned_ozone.csv', index=False)

> Till now **cleaning process** is completed, next we go for **Exploratory Data Analysis (EDA)**. Let's find out what happening in *skies* at **California**.

# Exploratory skies

In [ ]:
data = pd.read_csv('/content/drive/MyDrive/PROJECTS/data camp competition/cleaned_ozone.csv')
# fixing the date type, as csv stores dates as strings
data['Date'] = pd.to_datetime(data['Date'], errors='coerce')
data.head()

Date Source   Site ID  POC  Daily Max 8-hour Ozone Concentration  \
0 2024-01-01    AQS  60010007    1                                0.0310   
1 2024-01-02    AQS  60010007    1                                0.0370   
2 2024-01-03    AQS  60010007    1                                0.0864   
3 2024-01-04    AQS  60010007    1                                0.0260   
4 2024-01-05    AQS  60010007    1                                0.0270   

  Units  Daily AQI Value Local Site Name  Daily Obs Count  Percent Complete  \
0   ppm             29.0       Livermore               17             100.0   
1   ppm             34.0       Livermore               17             100.0   
2   ppm             30.0       Livermore               17             100.0   
3   ppm             24.0       Livermore               17             100.0   
4   ppm             25.0       Livermore               17             100.0   

   Method Code  CBSA Code                          CBSA Name  \
0           47      41860  San Francisco-Oakland-Hayward, CA   
1           47      41860  San Francisco-Oakland-Hayward, CA   
2           47      41860  San Francisco-Oakland-Hayward, CA   
3           47      41860  San Francisco-Oakland-Hayward, CA   
4           47      41860  San Francisco-Oakland-Hayward, CA   

   County FIPS Code   County  Site Latitude  Site Longitude  
0                 1  Alameda      37.687526     -121.784217  
1                 1  Alameda      37.687526     -121.784217  
2                 1  Alameda      37.687526     -121.784217  
3                 1  Alameda      37.687526     -121.784217  
4                 1  Alameda      37.687526     -121.784217

In [ ]:
data.isna().sum()

Date                                    0
Source                                  0
Site ID                                 0
POC                                     0
Daily Max 8-hour Ozone Concentration    0
Units                                   0
Daily AQI Value                         0
Local Site Name                         0
Daily Obs Count                         0
Percent Complete                        0
Method Code                             0
CBSA Code                               0
CBSA Name                               0
County FIPS Code                        0
County                                  0
Site Latitude                           0
Site Longitude                          0
dtype: int64

In [ ]:
# duplicate check
data.duplicated().sum()

np.int64(0)